File: lab5c#/Device.cs
using System;
using System.Text;

namespace SmartHomeSystem
{
    public class Device
    {
        // Статическое поле для подсчета всех устройств
        private static int totalDeviceCount = 0;

        // Статическое свойство для доступа
        public static int TotalDeviceCount => totalDeviceCount;

        // Статический метод для получения информации
        public static string GetDeviceStatistics()
        {
            return $"Всего устройств в системе: {totalDeviceCount}";
        }

        private string name;
        private double powerConsumption;

        public string DeviceId { get; private set; }
        public string Name
        {
            get => name;
            set
            {
                if (string.IsNullOrWhiteSpace(value))
                    throw new ArgumentException("Название устройства не может быть пустым", nameof(Name));

                if (value.Length > 50)
                    throw new ArgumentException("Название устройства слишком длинное (макс. 50 символов)", nameof(Name));

                name = value;
            }
        }
        public string Manufacturer { get; set; }
        public DeviceType DeviceType { get; set; }
        public Room Location { get; set; }
        public bool IsOn { get; private set; }
        public bool IsOnline { get; set; } = true;
        public double PowerConsumption
        {
            get => powerConsumption;
            set
            {
                if (value < 0)
                    throw new ArgumentException("Потребляемая мощность не может быть отрицательной", nameof(PowerConsumption));

                if (value > 10000)
                    throw new ArgumentException("Слишком высокая мощность (максимум 10000 Вт)", nameof(PowerConsumption));

                powerConsumption = value;
            }
        }

        public Device(string deviceId, string name, string manufacturer,
                     DeviceType deviceType, Room location, double powerConsumption)
        {
            if (string.IsNullOrWhiteSpace(deviceId))
                throw new ArgumentException("ID устройства не может быть пустым", nameof(deviceId));

            DeviceId = deviceId;
            Name = name;
            Manufacturer = manufacturer;
            DeviceType = deviceType;
            Location = location;
            PowerConsumption = powerConsumption;
            IsOn = false;

            totalDeviceCount++;  // Увеличиваем счетчик при создании устройства
        }

        // Деструктор для уменьшения счетчика
        ~Device()
        {
            totalDeviceCount--;
        }

        public void TurnOn()
        {
            IsOn = true;
            Console.WriteLine($"[+] Устройство '{Name}' включено.");
        }

        public void TurnOff()
        {
            IsOn = false;
            Console.WriteLine($"[−] Устройство '{Name}' выключено.");
        }

        public void UpdatePowerConsumption(double value)
        {
            try
            {
                PowerConsumption = value;
                Console.WriteLine($"Мощность устройства '{Name}' изменена на {value} Вт");
            }
            catch (ArgumentException ex)
            {
                Console.WriteLine($"Ошибка обновления мощности: {ex.Message}");
                throw;
            }
        }

        public string GetDeviceTypeString()
        {
            switch (DeviceType)
            {
                case DeviceType.SENSOR: return "Датчик";
                case DeviceType.ACTUATOR: return "Исполнительное";
                case DeviceType.CLIMATE_CONTROL: return "Климат";
                case DeviceType.SECURITY: return "Безопасность";
                case DeviceType.MULTIMEDIA: return "Мультимедиа";
                default: return "Неизвестно";
            }
        }

        public string GetStatus()
        {
            return IsOn ? "ВКЛ" : "ВЫКЛ";
        }

        public string Serialize()
        {
            return $"{DeviceId}|{Name}|{Manufacturer}|{(int)DeviceType}|" +
                   $"{(Location != null ? Location.RoomId : "NULL")}|{PowerConsumption}";
        }

        public static Device Deserialize(string data, Room location)
        {
            if (string.IsNullOrWhiteSpace(data))
                throw new ArgumentException("Данные для десериализации не могут быть пустыми", nameof(data));

            var parts = data.Split('|');
            if (parts.Length != 6)
                throw new FormatException($"Неверный формат данных устройства: {data}");

            try
            {
                return new Device(
                    parts[0],
                    parts[1],
                    parts[2],
                    (DeviceType)int.Parse(parts[3]),
                    location,
                    double.Parse(parts[5])
                );
            }
            catch (FormatException ex)
            {
                throw new FormatException($"Ошибка парсинга данных устройства: {data}", ex);
            }
            catch (ArgumentException ex)
            {
                throw new FormatException($"Ошибка валидации данных устройства: {data}", ex);
            }
        }

        public override string ToString()
        {
            return $"{Name} ({GetDeviceTypeString()}) - {GetStatus()}";
        }
    }
}

File: Labr3/device.cpp
#include "device.hpp"
#include "room.hpp"
#include <sstream>

Device::Device(const std::string& id, const std::string& deviceName,
    const std::string& manuf, DeviceType type, Room* room, double power)
    : deviceId(id), name(deviceName), manufacturer(manuf),
    deviceType(type), location(room), isOn(false), isOnline(true), powerConsumption(power) {}

Device::~Device() {}

void Device::turnOn() {
    isOn = true;
}

void Device::turnOff() {
    isOn = false;
}

std::string Device::getStatus() const {
    return isOn ? "ВКЛ" : "ВЫКЛ";
}

void Device::updateCurrentValue(double value) {
    powerConsumption = value;
}

std::string Device::getDeviceTypeString() const {
    switch (deviceType) {
    case DeviceType::SENSOR: return "Датчик";
    case DeviceType::ACTUATOR: return "Исполнительное устройство";
    case DeviceType::CLIMATE_CONTROL: return "Климат-контроль";
    case DeviceType::SECURITY: return "Безопасность";
    case DeviceType::MULTIMEDIA: return "Мультимедиа";
    default: return "Неизвестно";
    }
}

std::string Device::getDeviceId() const {
    return deviceId;
}

std::string Device::getName() const {
    return name;
}

std::string Device::getManufacturer() const {
    return manufacturer;
}

DeviceType Device::getDeviceType() const {
    return deviceType;
}

Room* Device::getLocation() const {
    return location;
}

bool Device::getIsOn() const {
    return isOn;
}

bool Device::getIsOnline() const {
    return isOnline;
}

double Device::getPowerConsumption() const {
    return powerConsumption;
}

void Device::setPowerConsumption(double power) {
    powerConsumption = power;
}

void Device::setIsOnline(bool online) {
    isOnline = online;
}

void Device::setLocation(Room* newLocation) {
    location = newLocation;
}

std::string Device::serialize() const {
    std::stringstream ss;
    ss << deviceId << "|" << name << "|" << manufacturer << "|"
        << static_cast<int>(deviceType) << "|" << (location ? location->getRoomId() : "NULL") << "|"
        << isOn << "|" << isOnline << "|" << powerConsumption;
    return ss.str();
}

Device* Device::deserialize(const std::string& data, Room* location) {
    std::stringstream ss(data);
    std::string id, name, manufacturer, typeStr, locationId, isOnStr, isOnlineStr, powerStr;

    std::getline(ss, id, '|');
    std::getline(ss, name, '|');
    std::getline(ss, manufacturer, '|');
    std::getline(ss, typeStr, '|');
    std::getline(ss, locationId, '|');
    std::getline(ss, isOnStr, '|');
    std::getline(ss, isOnlineStr, '|');
    std::getline(ss, powerStr, '|');

    DeviceType type = static_cast<DeviceType>(std::stoi(typeStr));
    Device* device = new Device(id, name, manufacturer, type, location, std::stod(powerStr));
    device->setIsOnline(isOnlineStr == "1");
    if (isOnStr == "1") device->turnOn();

    return device;
}